tests/machine_build.rs

```diff
@@ -5,17 +5,13 @@ use ckb_vm::machine::asm::{AsmCoreMachine, AsmMachine};
 use ckb_vm::machine::{trace::TraceMachine, DefaultCoreMachine, VERSION1, VERSION2};
 use ckb_vm::registers::{A0, A7};
 use ckb_vm::{
-    DefaultMachineBuilder, Error, Register, SparseMemory, SupportMachine, Syscalls, WXorXMemory,
-    ISA_A, ISA_B, ISA_IMC, ISA_MOP,
+    DefaultMachineBuilder, Error, ExecutionContext, Register, SparseMemory, SupportMachine,
+    WXorXMemory, ISA_A, ISA_B, ISA_IMC, ISA_MOP,
 };
 
-pub struct SleepSyscall {}
-
-impl<Mac: SupportMachine> Syscalls<Mac> for SleepSyscall {
-    fn initialize(&mut self, _machine: &mut Mac) -> Result<(), Error> {
-        Ok(())
-    }
+pub struct SleepContext;
 
+impl<Mac: SupportMachine> ExecutionContext<Mac> for SleepContext {
```

Comment:
Here we are still requiring syscalls to implement ExecutionContext, which is quite some unnecesary change for the API side. I would suggest we have Syscalls implement ExecutionContext. If as you mentioned, one needs to distinguish between ExecutionContext, and `Box<dyn>` , we should let them distinguish between a `Syscalls` and `Box<dyn>` or slightly rename them if you like.

Reply:
This is when you need to name the context type or machine type. If you don't expand `SleepSyscall` to `SleepContext` you'd need to use `WithCyclesFunc<..., fn (Instruction) -> u64>` or `Box<dyn>` or impl trait in return type, which are all possible but not really pretty.

Reply:
It's possible for T: Syscalls -> T: ExecutionContext, but a cycles func is usually also used when there's custom syscalls. It can be added nevertheless.

Reply:
This is actually not possible: conflicting implementations of trait `context::ExecutionContext<_>` for type `Box<(dyn context::ExecutionContext<_> + Send + Sync + 'static)>`
downstream crates may implement trait `syscalls::Syscalls<_>` for type `std::boxed::Box<(dyn context::ExecutionContext<_> + std::marker::Send + std::marker::Sync + 'static)>`